clang prints a lot of test build warnings

```
test_xoptional.cpp:203:34: warning: implicit conversion turns floating-point number into integer: 'double' to 'bool' [-Wfloat-conversion]
        EXPECT_EQ(res5, optional(d1 || d2, true));
                                 ^~ ~~
/usr/local/include/gtest/gtest.h:1969:29: note: expanded from macro 'EXPECT_EQ'
                      val1, val2)
                            ^~~~
/usr/local/include/gtest/gtest_pred_impl.h:161:40: note: expanded from macro 'EXPECT_PRED_FORMAT2'
  GTEST_PRED_FORMAT2_(pred_format, v1, v2, GTEST_NONFATAL_FAILURE_)
                                       ^~
/usr/local/include/gtest/gtest_pred_impl.h:146:43: note: expanded from macro 'GTEST_PRED_FORMAT2_'
  GTEST_ASSERT_(pred_format(#v1, #v2, v1, v2), \
                                          ^~
/usr/local/include/gtest/gtest_pred_impl.h:76:52: note: expanded from macro 'GTEST_ASSERT_'
  if (const ::testing::AssertionResult gtest_ar = (expression)) \
                                                   ^~~~~~~~~~
test_xoptional.cpp:203:40: warning: implicit conversion turns floating-point number into integer: 'double' to 'bool' [-Wfloat-conversion]
        EXPECT_EQ(res5, optional(d1 || d2, true));
                                    ~~ ^~
/usr/local/include/gtest/gtest.h:1969:29: note: expanded from macro 'EXPECT_EQ'
                      val1, val2)
                            ^~~~
/usr/local/include/gtest/gtest_pred_impl.h:161:40: note: expanded from macro 'EXPECT_PRED_FORMAT2'
  GTEST_PRED_FORMAT2_(pred_format, v1, v2, GTEST_NONFATAL_FAILURE_)
                                       ^~
/usr/local/include/gtest/gtest_pred_impl.h:146:43: note: expanded from macro 'GTEST_PRED_FORMAT2_'
  GTEST_ASSERT_(pred_format(#v1, #v2, v1, v2), \
                                          ^~
/usr/local/include/gtest/gtest_pred_impl.h:76:52: note: expanded from macro 'GTEST_ASSERT_'
  if (const ::testing::AssertionResult gtest_ar = (expression)) \
                                                   ^~~~~~~~~~
test_xoptional.cpp:206:34: warning: implicit conversion turns floating-point number into integer: 'double' to 'bool' [-Wfloat-conversion]
        EXPECT_EQ(res6, optional(d1 && d2, true));
                                 ^~ ~~
/usr/local/include/gtest/gtest.h:1969:29: note: expanded from macro 'EXPECT_EQ'
                      val1, val2)
                            ^~~~
/usr/local/include/gtest/gtest_pred_impl.h:161:40: note: expanded from macro 'EXPECT_PRED_FORMAT2'
  GTEST_PRED_FORMAT2_(pred_format, v1, v2, GTEST_NONFATAL_FAILURE_)
                                       ^~
/usr/local/include/gtest/gtest_pred_impl.h:146:43: note: expanded from macro 'GTEST_PRED_FORMAT2_'
  GTEST_ASSERT_(pred_format(#v1, #v2, v1, v2), \
                                          ^~
/usr/local/include/gtest/gtest_pred_impl.h:76:52: note: expanded from macro 'GTEST_ASSERT_'
  if (const ::testing::AssertionResult gtest_ar = (expression)) \
                                                   ^~~~~~~~~~
test_xoptional.cpp:206:40: warning: implicit conversion turns floating-point number into integer: 'double' to 'bool' [-Wfloat-conversion]
        EXPECT_EQ(res6, optional(d1 && d2, true));
                                    ~~ ^~
/usr/local/include/gtest/gtest.h:1969:29: note: expanded from macro 'EXPECT_EQ'
                      val1, val2)
                            ^~~~
/usr/local/include/gtest/gtest_pred_impl.h:161:40: note: expanded from macro 'EXPECT_PRED_FORMAT2'
  GTEST_PRED_FORMAT2_(pred_format, v1, v2, GTEST_NONFATAL_FAILURE_)
                                       ^~
/usr/local/include/gtest/gtest_pred_impl.h:146:43: note: expanded from macro 'GTEST_PRED_FORMAT2_'
  GTEST_ASSERT_(pred_format(#v1, #v2, v1, v2), \
                                          ^~
/usr/local/include/gtest/gtest_pred_impl.h:76:52: note: expanded from macro 'GTEST_ASSERT_'
  if (const ::testing::AssertionResult gtest_ar = (expression)) \
                                                   ^~~~~~~~~~
4 warnings generated.
[26/52] /usr/bin/c++  -DHAVE_NLOHMANN_JSON -isystem /usr/local/include -isystem /usr/ports/devel/xtl-quant-stack/work/stage/usr/local/include -O2 -pipe -fno-omit-frame-pointer -fstack-protector -fno-strict-aliasing -fno-omit-frame-pointer -O2 -pipe -fno-omit-frame-pointer -fstack-protector -fno-strict-aliasing -fno-omit-frame-pointer   -Wunused-parameter -Wextra -Wreorder -Wconversion -Wsign-conversion -march=native -pthread -std=gnu++14 -MD -MT CMakeFiles/test_xcomplex.dir/test_xcomplex.cpp.o -MF CMakeFiles/test_xcomplex.dir/test_xcomplex.cpp.o.d -o CMakeFiles/test_xcomplex.dir/test_xcomplex.cpp.o -c test_xcomplex.cpp
[27/52] /usr/bin/c++  -DHAVE_NLOHMANN_JSON -isystem /usr/local/include -isystem /usr/ports/devel/xtl-quant-stack/work/stage/usr/local/include -O2 -pipe -fno-omit-frame-pointer -fstack-protector -fno-strict-aliasing -fno-omit-frame-pointer -O2 -pipe -fno-omit-frame-pointer -fstack-protector -fno-strict-aliasing -fno-omit-frame-pointer   -Wunused-parameter -Wextra -Wreorder -Wconversion -Wsign-conversion -march=native -pthread -std=gnu++14 -MD -MT CMakeFiles/test_xmasked_value.dir/test_xmasked_value.cpp.o -MF CMakeFiles/test_xmasked_value.dir/test_xmasked_value.cpp.o.d -o CMakeFiles/test_xmasked_value.dir/test_xmasked_value.cpp.o -c test_xmasked_value.cpp
[28/52] : && /usr/bin/c++  -O2 -pipe -fno-omit-frame-pointer -fstack-protector -fno-strict-aliasing -fno-omit-frame-pointer -O2 -pipe -fno-omit-frame-pointer -fstack-protector -fno-strict-aliasing -fno-omit-frame-pointer  -fstack-protector CMakeFiles/test_xoptional.dir/test_xoptional.cpp.o  -o test_xoptional  -Wl,-rpath,/usr/local/lib /usr/local/lib/libgtest.so /usr/local/lib/libgtest_main.so -pthread && :
[29/52] : && /usr/bin/c++  -O2 -pipe -fno-omit-frame-pointer -fstack-protector -fno-strict-aliasing -fno-omit-frame-pointer -O2 -pipe -fno-omit-frame-pointer -fstack-protector -fno-strict-aliasing -fno-omit-frame-pointer  -fstack-protector CMakeFiles/test_xcomplex.dir/test_xcomplex.cpp.o  -o test_xcomplex  -Wl,-rpath,/usr/local/lib /usr/local/lib/libgtest.so /usr/local/lib/libgtest_main.so -pthread && :
[30/52] : && /usr/bin/c++  -O2 -pipe -fno-omit-frame-pointer -fstack-protector -fno-strict-aliasing -fno-omit-frame-pointer -O2 -pipe -fno-omit-frame-pointer -fstack-protector -fno-strict-aliasing -fno-omit-frame-pointer  -fstack-protector CMakeFiles/test_xmasked_value.dir/test_xmasked_value.cpp.o  -o test_xmasked_value  -Wl,-rpath,/usr/local

test_xhash.cpp:25:49: warning: implicit conversion changes signedness: 'int' to 'unsigned long' [-Wsign-conversion]
        uint8_t* hashes = new uint8_t[hashbytes * 256];
                          ~~~         ~~~~~~~~~~^~~~~
test_xhash.cpp:26:36: warning: implicit conversion changes signedness: 'int' to 'unsigned long' [-Wsign-conversion]
        uint8_t* res = new uint8_t[hashbytes];
                       ~~~         ^~~~~~~~~
test_xhash.cpp:94:41: warning: implicit conversion changes signedness: 'int' to 'unsigned long' [-Wsign-conversion]
        uint8_t * buffer1 = new uint8_t[buflen];
                            ~~~         ^~~~~~
test_xhash.cpp:95:41: warning: implicit conversion changes signedness: 'int' to 'unsigned long' [-Wsign-conversion]
        uint8_t * buffer2 = new uint8_t[buflen];
                            ~~~         ^~~~~~
test_xhash.cpp:97:39: warning: implicit conversion changes signedness: 'int' to 'unsigned long' [-Wsign-conversion]
        uint8_t * hash1 = new uint8_t[hashbytes];
                          ~~~         ^~~~~~~~~
test_xhash.cpp:98:39: warning: implicit conversion changes signedness: 'int' to 'unsigned long' [-Wsign-conversion]
        uint8_t * hash2 = new uint8_t[hashbytes];
                          ~~~         ^~~~~~~~~
test_xhash.cpp:25:49: warning: implicit conversion changes signedness: 'int' to 'unsigned long' [-Wsign-conversion]
        uint8_t* hashes = new uint8_t[hashbytes * 256];
                          ~~~         ~~~~~~~~~~^~~~~
test_xhash.cpp:159:27: note: in instantiation of function template specialization 'xtl::verification_test<unsigned long (*)(const void *, unsigned long, unsigned long)>' requested here
        uint32_t actual = verification_test(&hash_bytes, sizeof(std::size_t));
                          ^
test_xhash.cpp:26:36: warning: implicit conversion changes signedness: 'int' to 'unsigned long' [-Wsign-conversion]
        uint8_t* res = new uint8_t[hashbytes];
                       ~~~         ^~~~~~~~~
test_xhash.cpp:94:41: warning: implicit conversion changes signedness: 'int' to 'unsigned long' [-Wsign-conversion]
        uint8_t * buffer1 = new uint8_t[buflen];
                            ~~~         ^~~~~~
test_xhash.cpp:166:21: note: in instantiation of function template specialization 'xtl::sanity_test<unsigned long (*)(const void *, unsigned long, unsigned long)>' requested here
        EXPECT_TRUE(sanity_test(&hash_bytes, sizeof(std::size_t)));
                    ^
test_xhash.cpp:95:41: warning: implicit conversion changes signedness: 'int' to 'unsigned long' [-Wsign-conversion]
        uint8_t * buffer2 = new uint8_t[buflen];
                            ~~~         ^~~~~~
test_xhash.cpp:97:39: warning: implicit conversion changes signedness: 'int' to 'unsigned long' [-Wsign-conversion]
        uint8_t * hash1 = new uint8_t[hashbytes];
                          ~~~         ^~~~~~~~~
test_xhash.cpp:98:39: warning: implicit conversion changes signedness: 'int' to 'unsigned long' [-Wsign-conversion]
        uint8_t * hash2 = new uint8_t[hashbytes];
                          ~~~         ^~~~~~~~~
12 warnings generated.

```

--- a/include/xtl/xbasic_fixed_string.hpp
+++ b/include/xtl/xbasic_fixed_string.hpp
@@ -111,7 +111,7 @@ namespace xtl
         {
             fixed_string_external_storage_impl() = default;
 
-            fixed_string_external_storage_impl(T ptr, std::ptrdiff_t size)
+            fixed_string_external_storage_impl(T ptr, std::ptrdiff_t/*size*/)
             {
                 m_buffer = ptr;
             }

--- a/include/xtl/xhash.hpp
+++ b/include/xtl/xhash.hpp
@@ -34,7 +34,7 @@ namespace xtl
             const char* data = static_cast<const char*>(buffer);
             for (; length != 0; --length)
             {
-                hash = (hash * 131) + *data++;
+                hash = (hash * 131) + static_cast<std::size_t>(*data++);
             }
             return hash;
         }

--- a/include/xtl/xspan_impl.hpp
+++ b/include/xtl/xspan_impl.hpp
@@ -298,7 +298,7 @@ public:
     using reverse_iterator = std::reverse_iterator<iterator>;
     using const_reverse_iterator = std::reverse_iterator<const_iterator>;
 
-    static constexpr index_type extent = Extent;
+    static constexpr index_type extent = static_cast<index_type>(Extent);
 
     // [span.cons], span constructors, copy, assignment, and destructor
     template <std::ptrdiff_t E = Extent,
@@ -445,7 +445,7 @@ public:
     }
 
     TCB_SPAN_CONSTEXPR11 span<element_type, dynamic_extent>
-    subspan(index_type offset, index_type count = dynamic_extent) const
+    subspan(index_type offset, index_type count = static_cast<index_type>(dynamic_extent)) const
     {
         TCB_SPAN_EXPECT((offset >= 0 && offset <= size()) &&
                         (count == dynamic_extent ||

--- a/test/test_xbase64.cpp
+++ b/test/test_xbase64.cpp
@@ -16,24 +16,24 @@ namespace xtl
 {
     TEST(xbase64, encode)
     {
-       base64encode("") == "";
-       base64encode("f") == "Zg==";
-       base64encode("fo") == "Zm8=";
-       base64encode("foo") == "Zm9v";
-       base64encode("foob") == "Zm9vYg==";
-       base64encode("fooba") == "Zm9vYmE=";
-       base64encode("foobar") == "Zm9vYmFy";
+       EXPECT_EQ(base64encode(""), "");
+       EXPECT_EQ(base64encode("f"), "Zg==");
+       EXPECT_EQ(base64encode("fo"), "Zm8=");
+       EXPECT_EQ(base64encode("foo"), "Zm9v");
+       EXPECT_EQ(base64encode("foob"), "Zm9vYg==");
+       EXPECT_EQ(base64encode("fooba"), "Zm9vYmE=");
+       EXPECT_EQ(base64encode("foobar"), "Zm9vYmFy");
     }
 
     TEST(xbase64, decode)
     {
-       base64decode("") == "";
-       base64decode("f") == "Zg==";
-       base64decode("Zm8=") == "fo";
-       base64decode("Zm9v") == "foo";
-       base64decode("Zm9vYg==") == "foob";
-       base64decode("Zm9vYmE=") == "fooba";
-       base64decode("Zm9vYmFy") == "foobar";
+       EXPECT_EQ(base64decode(""), "");
+       EXPECT_EQ(base64decode("Zg=="), "f");
+       EXPECT_EQ(base64decode("Zm8="), "fo");
+       EXPECT_EQ(base64decode("Zm9v"), "foo");
+       EXPECT_EQ(base64decode("Zm9vYg=="), "foob");
+       EXPECT_EQ(base64decode("Zm9vYmE="), "fooba");
+       EXPECT_EQ(base64decode("Zm9vYmFy"), "foobar");
     }
 }
 

--- a/test/test_xhash.cpp
+++ b/test/test_xhash.cpp
@@ -19,26 +19,26 @@ namespace xtl
 
     // Adaptation of tests provided by Austin Appleby in https://github.com/aappleby/smhasher
     template <class F>
-    uint32_t verification_test(F f, int hashbytes)
+    uint32_t verification_test(F f, std::size_t hashbytes)
     {
         uint8_t* key = new uint8_t[256];
         uint8_t* hashes = new uint8_t[hashbytes * 256];
         uint8_t* res = new uint8_t[hashbytes];
 
         std::memset(key, 0, 256);
-        std::memset(hashes, 0, std::size_t(hashbytes * 256));
-        std::memset(res, 0, std::size_t(hashbytes));
+        std::memset(hashes, 0, hashbytes * 256);
+        std::memset(res, 0, hashbytes);
 
         // Hash keys of the form {0}, {0,1}, {0,1,2}... up to N=255,using 256-N as
         // the seed
         for(std::size_t i = 0; i < 256; i++)
         {
             key[i] = (uint8_t)i;
-            *reinterpret_cast<std::size_t*>(hashes + int(i) * hashbytes) = f(key,i,256-i);
+            *reinterpret_cast<std::size_t*>(hashes + i * hashbytes) = f(key,i,256-i);
         }
 
         // Then hash the result array
-        *reinterpret_cast<std::size_t*>(res) = f(hashes,std::size_t(hashbytes)*256,0);
+        *reinterpret_cast<std::size_t*>(res) = f(hashes,hashbytes*256,0);
 
         // The first four bytes of that hash, interpreted as a little-endian integer, is our
         // verification value
@@ -81,7 +81,7 @@ namespace xtl
     }
 
     template <class F>
-    bool sanity_test(F f, int hashbytes)
+    bool sanity_test(F f, std::size_t hashbytes)
     {
         std::srand(883741);
         bool result = true;
@@ -91,8 +91,8 @@ namespace xtl
         int pad = 16;
         int buflen = keymax + pad*3;
   
-        uint8_t * buffer1 = new uint8_t[buflen];
-        uint8_t * buffer2 = new uint8_t[buflen];
+        uint8_t * buffer1 = new uint8_t[std::size_t(buflen)];
+        uint8_t * buffer2 = new uint8_t[std::size_t(buflen)];
 
         uint8_t * hash1 = new uint8_t[hashbytes];
         uint8_t * hash2 = new uint8_t[hashbytes];
@@ -119,7 +119,7 @@ namespace xtl
                         flipbit(key2,len,uint32_t(bit));
                         *reinterpret_cast<std::size_t*>(hash2) = f(key2,std::size_t(len),0);
 
-                        if(std::memcmp(hash1,hash2,std::size_t(hashbytes)) == 0)
+                        if(std::memcmp(hash1,hash2,hashbytes) == 0)
                         {
                             result = false;
                         }
@@ -127,7 +127,7 @@ namespace xtl
                         flipbit(key2,len,uint32_t(bit));
                         *reinterpret_cast<std::size_t*>(hash2) = f(key2,std::size_t(len),0);
 
-                        if(std::memcmp(hash1,hash2,std::size_t(hashbytes)) != 0)
+                        if(std::memcmp(hash1,hash2,hashbytes) != 0)
                         {
                             result = false;
                         }

--- a/test/test_xoptional.cpp
+++ b/test/test_xoptional.cpp
@@ -199,12 +199,6 @@ namespace xtl
         auto res4 = o1 / o2;
         EXPECT_EQ(res4, optional(d1 / d2, true));
 
-        auto res5 = o1 || o2;
-        EXPECT_EQ(res5, optional(d1 || d2, true));
-
-        auto res6 = o1 && o2;
-        EXPECT_EQ(res6, optional(d1 && d2, true));
-
         xoptional<bool> res7 = o1 < o2;
         EXPECT_TRUE(res7.value());
 
@@ -237,6 +231,12 @@ namespace xtl
 
         auto res13 = ~oi1;
         EXPECT_EQ(res13, optional(~i1, true));
+        
+        auto res5 = oi1 || oi2;
+        EXPECT_EQ(res5, optional(i1 || i2, true));
+
+        auto res6 = oi1 && oi2;
+        EXPECT_EQ(res6, optional(i1 && i2, true));
     }
 
     TEST(xoptional, free_functions)

